src/rust/engine/fs/src/directory.rs

```diff
@@ -0,0 +1,388 @@
+  type Target = Intern<String>;
+
+  fn deref(&self) -> &Intern<String> {
+    &self.0
+  }
+}
+
+#[derive(DeepSizeOf)]
+pub enum Entry {
+  Directory(Directory),
+  File(File),
+}
+
+impl Entry {
+  fn name(&self) -> Name {
+    match self {
+      Entry::Directory(d) => d.name,
+      Entry::File(f) => f.name,
+    }
+  }
+}
+
+#[derive(DeepSizeOf)]
+pub struct Directory {
+  name: Name,
+  digest: Digest,
+  tree: DigestTrie,
+}
+
+impl Directory {
+  fn new(name: Name, entries: Vec<Entry>) -> Self {
+    Self::from_digest_tree(name, DigestTrie(entries.into()))
+  }
+
+  fn from_digest_tree(name: Name, tree: DigestTrie) -> Self {
+    Self {
+      name,
+      digest: tree.compute_root_digest(),
+      tree,
+    }
+  }
+
+  pub fn as_directory(&self) -> remexec::Directory {
+    self.tree.as_directory()
+  }
+
+  pub fn as_directory_node(&self) -> remexec::DirectoryNode {
+    remexec::DirectoryNode {
+      name: self.name.as_ref().to_owned(),
+      digest: Some((&self.digest).into()),
+    }
+  }
+}
+
+#[derive(DeepSizeOf)]
+pub struct File {
+  name: Name,
+  digest: Digest,
+  is_executable: bool,
+}
+
+impl File {
+  pub fn as_file_node(&self) -> remexec::FileNode {
+    remexec::FileNode {
+      name: self.name.as_ref().to_owned(),
+      digest: Some(self.digest.into()),
+      is_executable: self.is_executable,
+      ..remexec::FileNode::default()
+    }
+  }
+}
+
+// TODO: `PathStat` owns its path, which means it can't be used via recursive slicing. See
+// whether these types can be merged.
+enum TypedPath<'a> {
+  File { path: &'a Path, is_executable: bool },
+  Dir(&'a Path),
+}
+
+impl<'a> Deref for TypedPath<'a> {
+  type Target = Path;
+
+  fn deref(&self) -> &Path {
+    match self {
+      TypedPath::File { path, .. } => path,
+      TypedPath::Dir(d) => d,
+    }
+  }
+}
+
+impl<'a> From<&'a PathStat> for TypedPath<'a> {
+  fn from(p: &'a PathStat) -> Self {
+    match p {
+      PathStat::File { path, stat } => TypedPath::File {
+        path,
+        is_executable: stat.is_executable,
+      },
+      PathStat::Dir { path, .. } => TypedPath::Dir(path),
+    }
+  }
+}
+
+#[derive(Clone, DeepSizeOf)]
+pub struct DigestTrie(Arc<[Entry]>);
+
+// TODO: This avoids a `rustc` crasher (repro on 7f319ee84ad41bc0aea3cb01fb2f32dcd51be704).
+unsafe impl Sync for DigestTrie {}
+
+impl DigestTrie {
+  pub fn from_path_stats(
+    path_stats: Vec<PathStat>,
+    file_digests: &HashMap<PathBuf, Digest>,
+  ) -> Result<Self, String> {
+    let path_stats = PathStat::normalize_path_stats(path_stats)?;
+    Ok(Self::from_sorted_paths(
+      PathBuf::new(),
+      path_stats.iter().map(|p| p.into()).collect(),
+      file_digests,
+    ))
+  }
+
+  fn from_sorted_paths(
+    prefix: PathBuf,
+    paths: Vec<TypedPath>,
+    file_digests: &HashMap<PathBuf, Digest>,
+  ) -> Self {
+    let mut entries = Vec::new();
+
+    for (name, group) in &paths
+      .into_iter()
+      .group_by(|s| path_name_to_name(s).unwrap())
```

Comment:
Will this `.unwrap` never panic?

Reply:
Good call... this was hasty. Both of the failure cases for `path_name_to_name` are possible, though rare.